File: iOS/authservice/AGConnectAuthDemo/provider/FacebookProvider.m
/*
    Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

#import "FacebookProvider.h"

@implementation FacebookProvider
{
    FBSDKLoginManager *manager;
}

+ (instancetype)sharedInstance {
    static FacebookProvider * sharedInstance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    if (self = [super init]) {
        manager = [FBSDKLoginManager new];
    }
    return self;
}

- (void)startUpWithApp:(UIApplication *)application options:(NSDictionary *)launchOptions {
    [[FBSDKApplicationDelegate sharedInstance] application:application didFinishLaunchingWithOptions:launchOptions];
}

- (BOOL)application:(id)app openURL:(id)url options:(id)options {
    return [[FBSDKApplicationDelegate sharedInstance] application:app openURL:url sourceApplication:options[UIApplicationOpenURLOptionsSourceApplicationKey] annotation:options[UIApplicationOpenURLOptionsAnnotationKey] ];
}

- (void)fetchCredentialWithController:(UIViewController *)vc completion:(CredentialBlock)completion {
    self.credentialBlock = completion;
    [manager logInWithPermissions:@[@"public_profile",@"email"]
               fromViewController:vc
                          handler:^(FBSDKLoginManagerLoginResult * _Nullable result, NSError * _Nullable error) {
        if (!error) {
            AGCAuthCredential *credential = [AGCFacebookAuthProvider credentialWithToken:result.token.tokenString];
            if (self.credentialBlock) {
                self.credentialBlock(credential);
            }
        }
        
    }];
}

@end

File: iOS/authservice/AGConnectAuthDemo-Swift/provider/FacebookProvider.swift
/*
    Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

import UIKit
import AGConnectAuth

class FacebookProvider: BaseProvider {
    
    static let sharedInstance = FacebookProvider()
    
    let manager : LoginManager
    
    
    private override init() {
        manager = LoginManager()
    }
    
    func startUp(_ app:UIApplication, options:[UIApplication.LaunchOptionsKey : Any]) {
        ApplicationDelegate.shared.application(app, didFinishLaunchingWithOptions: options)
    }

    override func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        return ApplicationDelegate.shared.application(app, open: url, sourceApplication: options[UIApplication.OpenURLOptionsKey.sourceApplication] as? String, annotation: options[UIApplication.OpenURLOptionsKey.annotation])
    }

    override func fetchCredential(controller: UIViewController, completion: @escaping CredentialBlock) {
        credentialBlock = completion
        manager.logIn(permissions: ["public_profile","email"], from: controller) { (result, error) in
            if error==nil {
                let credential = AGCFacebookAuthProvider.credential(withToken: result?.token?.tokenString ?? "")
                if let credentialBlock = self.credentialBlock {
                    credentialBlock(credential)
                }
            }
        }
    }

}
